enterprise/backend/src/metabase_enterprise/representations/v0/common.clj

```diff
@@ -0,0 +1,251 @@
+(ns metabase-enterprise.representations.v0.common
+  (:require
+   [clojure.set :as set]
+   [clojure.string :as str]
+   [clojure.walk :as walk]
+   [metabase.util :as u]
+   [toucan2.core :as t2]))
+
+(set! *warn-on-reflection* true)
+
+(defmulti representation-type
+  "Returns the representation type (e.g., :question, :model, :metric) for a toucan entity.
+
+  Dispatches on the toucan model type."
+  {:arglists '[[entity]]}
+  t2/model)
+
+(defmethod representation-type :default [entity]
+  (throw (ex-info (str "Unknown entity type: " (t2/model entity))
+                  {:entity entity})))
+
+(defn entity-id
+  "Generates an entity-id stably from ref and collection-ref."
+  [ref collection-ref]
+  (-> (str collection-ref "/" ref)
+      hash
+      str
+      u/generate-nano-id))
+
+(defn generate-entity-id
+  "Generate a stable entity-id from the representation's collection-ref and its own ref."
+  [representation]
+  (entity-id (:name representation) (:collection representation)))
+
+(defn add-entity-id-from-ref-hash
+  "Add a stable entity-id to a representation based on its ref and collection ref."
+  [representation]
+  (assoc representation :entity-id (generate-entity-id representation)))
+
+(defn add-entity-id-random
+  "Add an unstable entity-id to a representation. It should be random so that it never collides."
+  [representation]
+  (assoc representation :entity-id (u/generate-nano-id)))
+
+(defn remove-entity-id
+  "Remove the entity id from a representation (if it has one)."
+  [representation]
+  (dissoc representation :entity-id))
+
+(defn find-collection-id
+  "Find collection ID by name or ref. Returns nil if not found."
+  [collection-ref]
+  (when collection-ref
+    (or
+     (when (integer? collection-ref) collection-ref)
+     ;; Try to find by slug or name
+     (t2/select-one-pk :model/Collection :slug collection-ref)
+     (t2/select-one-pk :model/Collection :name collection-ref))))
+
+(defn ref?
+  "Is it a ref?"
+  [x]
+  (and (string? x)
+       (str/starts-with? x "ref:")))
+
+(defn unref
+  "Unmake that ref."
+  [x]
+  (when (ref? x)
+    ;; "ref:"
+    (subs x 4)))
+
+(defn refs
+  "Returns all refs present in the entity-map, recursively walking to discover them."
+  [representation]
+  (let [v (volatile! [])]
```

Comment:
If you wanted to _really_ make this fast you could do

```suggestion
  (let [v (volatile! (transient []))]
```

and then use `conj!` and what not